Editorial review: Document responsive iframe sizing (#44598)
mdn/content: files/en-us/web/api/window/requestresize/index.md

---
title: "Window: requestResize() method"
short-title: requestResize()
slug: Web/API/Window/requestResize
page-type: web-api-instance-method
browser-compat: api.Window.requestResize
---

{{APIRef}}

> The **`requestResize()`** method of the {{domxref("Window")}} interface updates the size information shared by an embedded document with its embedding parent, but only if the embedded page has opted in to sharing its size information.

Reviewer: I see it sometimes say "embedded document", "embedded page", or "embedded content". Are they intentionally differentiated?

Author: No, this is just sloppy inconsistency. I've gone through and used "document" in all places, except where it is inappropriate.

## Syntax

```js-nolint
requestResize()
```

### Parameters

None.

### Return value

None ({{jsxref("undefined")}}).

## Description

For security and privacy reasons, {{htmlelement("iframe")}} elements do not by default expose any information to the parent page about the size of the content they are embedding.

To enable responsive sizing of {{htmlelement("iframe")}} elements based on their content, the [`<meta name="responsive-embedded-sizing">`](/en-US/docs/Web/HTML/Reference/Elements/meta/name/responsive-embedded-sizing) tag can be included in an embedded document to opt it in to sharing its size information with the parent page. The {{cssxref("frame-sizing")}} CSS property can then be set on the `<iframe>` to cause it to adopt the same horizontal or vertical size as the embedded content's actual content size (termed the **internal layout intrinsic size** in the spec). This is useful for avoiding scrollbars on embedded content so that it fits more seamlessly with its embedder.

> To resize the `<iframe>` dynamically as the embedded content changes size, you can call the {{domxref("Window.requestResize()")}} method from the embedded page to make it report an updated size, typically from within the event handler that caused the content to change size. If the `<iframe>` is sized using `frame-sizing`, it will then update its size automatically so that it still neatly contains the embeded content.

Reviewer: How about mentioning that the size is updated on `DOMContentLoad` and `Load` events automatically, and this method provides additional timing? Is this redundant?

Author: This is a really good piece of information to include. I've added a new paragraph to the bottom of the description:

> The embedded document's layout size is automatically reported once when its {{domxref("Document.DOMContentLoaded_event", "DOMContentLoaded")}} event fires, and again when the {{domxref("Window")}} object's {{domxref("Window.load_event", "load")}} event fires. The `requestResize()` method is needed to report subsequent layout size changes.

> 

Reviewer: How about also describing it'd raise an error if the child didn't opt-in? Is this too much details?

Please see the [CSS spec for `requestResize`](https://drafts.csswg.org/css-sizing-4/#window-requestresize)  for the details of the errors.

Maybe also nice to mention that, if the parent doesn't opt-in (by the `frame-sizing` property), it doesn't cause an error, but the iframe will not be resized?

Author: Good call. I've added an "Exceptions" subsection to describe the error conditions and note the behavior when `frame-sizing` is not set.

To enable responsive sizing of `<iframe>` elements based on their content, the [`<meta name="responsive-embedded-sizing">`](/en-US/docs/Web/HTML/Reference/Elements/meta/name/responsive-embedded-sizing) tag can be included in an embedded document to opt it in to sharing its size information with the parent document. The {{cssxref("frame-sizing")}} CSS property can then be set on the `<iframe>` to cause it to adopt the same horizontal or vertical size as the embedded document's actual layout size (termed the **internal layout intrinsic size** in the spec). This is useful for avoiding scrollbars on `<iframe>` content so that it fits more seamlessly with its embedder.

To resize the `<iframe>` dynamically as the embedded document changes layout size, you can call the {{domxref("Window.requestResize()")}} method from the embedded document to make it report an updated size, typically from within the event handler that caused its content to change size. If the `<iframe>` is sized using `frame-sizing`, it will then update its size automatically so that it still neatly contains the embeded content.

> The embedded document's layout size is automatically reported once when its {{domxref("Document.DOMContentLoaded_event", "DOMContentLoaded")}} event fires, and again when the {{domxref("Window")}} object's {{domxref("Window.load_event", "load")}} event fires. The `requestResize()` method is needed to report subsequent layout size changes.

Reviewer 2: I like this, but  it should precede the paragraph above. This may require some playing with the second sentence to avoid repetition.

Author: Good call on this. I've reversed the order and tweaked the text so that the narrative flow still works.

## Examples

### Using `requestResize()`

Our [`requestResize()` demo](https://mdn.github.io/dom-examples/responsive-iframe-sizing/js-request-resize/) ([see source code](https://github.com/mdn/dom-examples/tree/main/responsive-iframe-sizing/js-request-resize)) demonstrates usage of the `requestResize()` method.

> #### HTML

Reviewer 2: I found the jumping between the two docs a bit unhelpful. I'd dump the normal pattern of the HTML/CSS sub headings and use Index and Frame as your sections (along with results). Up to you.

Author: I've removed the links from the top of the demo sections so that you are not immediately jumping to a different page. They are now only at the bottom of the section. 

I've also tried out using index and frame sections, and I think it works well. Let me know what you think.

```css
iframe {
  frame-sizing: content-block-size;
  border: 2px solid gray;
}
```

#### JavaScript

The script inside `frame.html` starts by grabbing a reference to the `<div>` element. It then sets `click` and `keydown` event listeners on the `<div>`, both of which run a custom function called `addParagraph()` when the event fires.

```js
const divElem = document.querySelector("div");
divElem.addEventListener("click", addParagraph);
window.addEventListener("keydown", addParagraph);
```

The `addParagraph()` function generates a new paragraph element and appends it to the end of the `<div>` as a child, increasing its height. It then calls `requestResize()` so that the new height is reported to the parent document.

```js
function addParagraph() {
  const para = document.createElement("p");
  para.textContent = "New content.";
  divElem.appendChild(para);
  window.requestResize();
}
```

#### Result

{{EmbedGHLiveSample("dom-examples/responsive-iframe-sizing/js-request-resize/", "100%", 300)}}

> Even though no explicit `height` has been set on the `<iframe>`, it is sized to the right height to exactly contain its embedded document, with no scroll bars. Try clicking on the `<div>` or focusing it and pressing a key on the keyboard. As a new paragraph is added to the `<div>`, the `<div>` grows in height, but the `<iframe>` also grows in height to match it.

Reviewer 2: This is true, but it isn't observable in MDN because the frame is embedded in the live example frame. 
So the instructions need to be inverted to say "load this in separate tab" and not show the embedded version in MDN.

Author: OK, I've updated the demo sections to remove the embedded demo and direct readers to open the demo in a new tab from the get-go.

## Syntax

```js-nolint
requestResize()
```

### Parameters

None.

### Return value

None ({{jsxref("undefined")}}).

### Exceptions

- `NotAllowedError` {{domxref("DOMException")}}
  - : Thrown if:
    - The `requestResize()` method was called from a top-level (non-embedded) document.
    - The embedding element is not an {{htmlelement("iframe")}}.
    - The embedded document has not opted in to sharing its layout size by including a [`<meta name="responsive-embedded-sizing">`](/en-US/docs/Web/HTML/Reference/Elements/meta/name/responsive-embedded-sizing) tag.

> [!NOTE]
> If the parent document doesn't set the {{cssxref("frame-sizing")}} CSS property on the embedding `<iframe>`, no exception is thrown, but the `<iframe>` will not be resized.

## Description

For security and privacy reasons, {{htmlelement("iframe")}} elements do not by default expose any information to the parent document about the size of the content in the document they are embedding.

> To enable responsive sizing of `<iframe>` elements based on their content, the [`<meta name="responsive-embedded-sizing">`](/en-US/docs/Web/HTML/Reference/Elements/meta/name/responsive-embedded-sizing) tag can be included in an embedded document to opt it in to sharing its size information with the parent document. The {{cssxref("frame-sizing")}} CSS property can then be set on the `<iframe>` to cause it to adopt the same horizontal or vertical size as the embedded document's actual layout size (termed the **internal layout intrinsic size** in the spec). This is useful for avoiding scrollbars on `<iframe>` content so that it fits more seamlessly into its embedder.

Reviewer 2: Re-reading this, `This is useful for avoiding scrollbars on `<iframe>` content so that it fits more seamlessly into its embedder.` feels a bit clunky. UP to you, but maybe as below (and the 4 other instances):

```suggestion
To enable responsive sizing of `<iframe>` elements based on their content, the [`<meta name="responsive-embedded-sizing">`](/en-US/docs/Web/HTML/Reference/Elements/meta/name/responsive-embedded-sizing) tag can be included in an embedded document to opt it in to sharing its size information with the parent document. The {{cssxref("frame-sizing")}} CSS property can then be set on the `<iframe>` to cause it to adopt the same horizontal or vertical size as the embedded document's actual layout size (termed the **internal layout intrinsic size** in the spec). This ensures that `<iframe>` content fits seamlessly into its embedder, avoiding unnecessary scrollbars.
```

Author: Sure, this sounds nicer. I'll use this everywhere, but also try to vary it a little on each page.
